### PR TITLE
Creates ADO.NET, ADO.NET ORM, and ADO.NET Entity Framework pages (take 2)

### LINKED ISSUE #79: Consolidate C#, .NET, and ASP pages into single .NET page (with or without subpages)
The information for avoiding SQL injection in .NET is currently fragmented over 3 different pages: [C#](http://bobby-tables.com/csharp), [.NET](http://bobby-tables.com/dotnet), and [ASP](http://bobby-tables.com/asp). It should really be consolidated into a single page (with or without subpages as appropriate), and should be structured as follows:

* Direct ADO.NET
  - Using the base classes in [**System.Data.Common**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common?view=netframework-4.7.2)
  - [Built-in providers](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers)
    - [SQL Server](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers#net-framework-data-provider-for-includessnoversionincludesssnoversion-mdmd-sqlclient)
    - [OLE DB](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers#net-framework-data-provider-for-ole-db)
    - [ODBC](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers#net-framework-data-provider-for-odbc)
  - Third-party providers
    - [Oracle](http://www.oracle.com/technetwork/topics/dotnet/index-085163.html)
    - Postgresql
    - [MySQL](https://dev.mysql.com/downloads/connector/net/6.10.html)
    - [SQLite](https://system.data.sqlite.org/index.html/doc/trunk/www/index.wiki)
* Entity Framework Core ?
* EF 6
  - [ExecuteSQLCommand](https://msdn.microsoft.com/en-us/library/system.data.entity.database.executesqlcommand(v=vs.113).aspx) and [ExecuteSQLCommandSync](https://msdn.microsoft.com/en-us/library/system.data.entity.database.executesqlcommandasync(v=vs.113).aspx) methods
  - [SqlQuery](https://msdn.microsoft.com/en-us/library/system.data.entity.database.sqlquery(v=vs.113).aspx) method
* Dapper
* Additional ORMs

This information is the same for any .NET language -- C#, VB.NET, F#, IronPython -- and for every .NET programming environment -- ASP.NET, WPF, console application, WinForms. At best it might be good to have examples for various languages on each subsection.

One important note: some providers support named parameters (e.g. the SQL Server provider); some providers support position parameters (e.g. the OLE DB provider).

### PR DESCRIPTION
**Pull request by zspitz**
Fixes #79

@petdance AFAICT the ADO.NET is complete, but the EF page could use some additional examples.

@mason-mcglothlin @kimmel @deanebarker Again, as previous contributors to these pages, could you look over this and see if I've missed anything?

### CODE REVIEW

File: s/adodotnet.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+There are a number of [data providers built-in to the .NET Framework](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers), for:
+
+* [SQL Server](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers#net-framework-data-provider-for-sql-server-sqlclient)
+* [OLE DB data sources](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers#net-framework-data-provider-for-ole-db) -- Access, Excel, text files, and others
+* [ODBC data sources](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers#net-framework-data-provider-for-odbc)
+* [Entity Framework Models](https://msdn.microsoft.com/library/49202ab9-ac98-4b4b-a05c-140e422bf527)
+* [SQL Server Compact Edition](https://msdn.microsoft.com/library/system.data.sqlserverce.aspx)
+* [Oracle](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/data-providers#net-framework-data-provider-for-oracle) (deprecated; use [ODP](https://www.oracle.com/technetwork/topics/dotnet/index-085163.html))
```
Comment (mason-mcglothlin) [inline s/adodotnet.md]: The built-in data provider for Oracle has [been deprecated](http://blogs.msdn.com/adonet/archive/2009/06/15/system-data-oracleclient-update.aspx). Instead, the [Managed Driver](https://www.nuget.org/packages/Oracle.ManagedDataAccess/) is the most direct replacement: like System.Data.OracleClient it doesn't require any other software to function.

Comment (zspitz) [reply s/adodotnet.md]: Do you mean that the deprecation needs to be better emphasized? Or even that the built-in provider shouldn't be mentioned at all?

**Comment by petdance** [comment]
Please note.  I am currently working on the PR that you have sent me.  Please don't make any more changes to this PR until I have said it's OK.  There are a lot of changes I have to make on my end because the Markdown processor I am using doesn't handle GitHub Markdown.  Not sure if I'm going to update all your markdown, or if I'm going to look at a new processor.  Either way, don't make any more changes.

Thanks for your work on this.  This is so much knowledge that I wouldn't have.

**Comment by zspitz** [comment]
@petdance Got it. I'm assuming it wouldn't have made a difference if everything was a single commit,, and not multiple commits? I'm asking because I want to turn my attention to resolving #80  If you decide to keep the current Markdown processor, where could I find out which subset of Markdown is supported?

**Comment by petdance** [comment]
I'm working on a branch and then I'm going to squash it all down back to a single commit when I get things worked out and merged back to dev.

It looks like @autarch's Markdent module is going to save our bacon here.

**Comment by zspitz** [comment]
@petdance Are you planning on using Markdent as the new Markdown processor? Or just to convert these pages to something compatible with the previous processor?

**Comment by petdance** [comment]
I just pushed 8169423 to `dev` which uses Markdent.  Plus we're now HTML5.

**Comment by petdance** [comment]
I need to know what we're going to call these new pages in the left sidebar.

You've deleted dotnet.md which was ".NET". and added adodotnet.md, adodotnet_ef.md, adodotnet_orm.md, and vbdotnet.md.  Those all need labels that I can stick in the sidebar so people can nav to them.  Or should we only be able to get to _orm.md and _ef.md from the main adodotnet.md page?

**Comment by petdance** [comment]
I've created a new working branch `pr86` that has your text and the start of my work to integrate them.  Don't bother making changes to it, but let's just keep discussing here on this PR.

If you install all the Perl modules you need you can run `make` on it and see how things come out.

Thanks,
Andy

**Comment by zspitz** [comment]
> I need to know what we're going to call these new pages in the left sidebar.

`vbdotnet.md` -> VB.NET
`adodotnet.md` -> ADO.NET
`adodotnet_ef.md` -> Entity Framework, with a link to ADO.NET and ADO.NET ORM pages

For now, I think that the `adodotnet_orm.md` page shouldn't have an independent label, as you suggest, but there should be a link in the `adodotnet.md` page.

**Comment by zspitz** [comment]
Something is broken in the link to the CommandText property. It currently reads:

    [`CommandText`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbcommand.commandtext?view=netframework-4.7.2# System_Data_IDbCommand_CommandText)

with a space after the `#`. The link could actually be shortened to:

    [`CommandText`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbcommand.commandtext)

---

Also, the section on the actual steps to avoid SQL injection is repeated twice: once at the beginning, and once at the [**Avoiding SQL injection in commands**](https://github.com/petdance/bobby-tables/blob/pr86/s/adodotnet.md#avoiding-sql-injection-in-commands) subsection. But the table describing the placeholder syntax has value. I think the duplicate parts of this subsection should be deleted, and the subsection renamed to **Placeholder syntax**.

**Comment by petdance** [comment]
Fixed the space in the URL: f4bdaaa

**Comment by petdance** [comment]
> adodotnet_ef.md -> Entity Framework, with a link to ADO.NET and ADO.NET ORM pages

I don't think that works, since everything in the sidebar is basically a language.  It's part of ADO.NET somehow, right?  "Entity Framework is Microsoft's primary ORM (object-relational mapper) built on top of ADO.NET." says the file.

**Comment by petdance** [comment]
I'm closing this PR.

I've pulled most of it manually to the `dev` branch.  Most of the changes above have been made, except:

* The duplicated SQL injection section didn't get removed, because I wasn't sure which one we should throw away, and

* The tables aren't rendering correctly.  The parser not be handling them correctly.

So this is closed, and published, and let's move forward from here.  Thanks very much for all your work on this.

**Comment by zspitz** [comment]
RE: tables, I think Markdown allows for embedded HTML, so the tables could be written with HTML instead of Markdown.

**Comment by petdance** [comment]
Or, do they really need to be tables?  Could they be something like

* this thing
    * example
* that thing
    * example

**Comment by zspitz** [comment]
The table on placeholder syntax could be written:

    * **Provider**: SQL Server, Entity SQL
          
          SELECT * FROM Students WHERE FirstName = @FirstName

    * **Provider**: OLE DB, ODBC
          
          SELECT * FROM Students WHERE FirstName = ?

    * **Provider**: Oracle

          SELECT * FROM Students WHERE FirstName = :FirstName


and would look like this:

* **Provider**: SQL Server, Entity SQL
          
      SELECT * FROM Students WHERE FirstName = @FirstName

* **Provider**: OLE DB, ODBC
          
      SELECT * FROM Students WHERE FirstName = ?

* **Provider**: Oracle

      SELECT * FROM Students WHERE FirstName = :FirstName

---

The table which maps the inheritance for some SQL-Server-specific classes to the abstract base classes is harder, but maybe something like this:

    For example, the ADO.NET data provider for connecting to SQL Server contains the following classes: in the  [`System.Data.SqlCient`](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient) namespace:

    * [**SqlCommand**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlcommand) inherits from [**DbCommand**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbcommand)
    * [**SqlParameter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlparameter) inherits from [**DbParameter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbparameter)
    * [**SqlDataReader**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqldatareader) inherits from [**DbDataReader**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdatareader)|
    * [**SqlDataAdapter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqldataadapter) inherits from [**DbDataAdapter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdataadapter)

and would render like this:

For example, the ADO.NET data provider for connecting to SQL Server contains the following classes in the  [`System.Data.SqlCient`](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient) namespace:

* [**SqlCommand**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlcommand) inherits from [**DbCommand**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbcommand)
* [**SqlParameter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlparameter) inherits from [**DbParameter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbparameter)
* [**SqlDataReader**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqldatareader) inherits from [**DbDataReader**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdatareader)|
* [**SqlDataAdapter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqldataadapter) inherits from [**DbDataAdapter**](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdataadapter)

**Comment by petdance** [comment]
That looks fine to me.  What do you thinK?

**Comment by zspitz** [comment]
> What do you thinK?

I think it's OK.

---

> I don't think that works, since everything in the sidebar is basically a language.

That's because in most cases a data access API is used by one language. ADO.NET (as well as #80 ) are language-independent APIs. In that case I think it makes sense to have a separate label for the data access API, and different language pages should point to the data access API page (like the `vbdotnet.md` page).

> It's part of ADO.NET somehow, right?

It's built on top of ADO.NET, and used to be a built-in part of the framework; but that is no longer the case -- it's acquired separately from the .NET Framework, using the NuGet package manager (like npm, but for .NET).

**Comment by zspitz** [comment]
Except that nobody calls it _ADO.NET Entity Framework_; it's always referred to as _Entity Framework_, _Entity Framework 6_ or _Entity Framework Core_.

---

It's not just tables that aren't rendered properly, but lists as well:

(Snapshot from GitHub)
> image

(Snapshot from bobby-tables.com)
> image

**Comment by zspitz** [comment]
The formatting for the Powershell example is not rendering properly:

(Snapshot from GitHub)
> image

(Snapshot from bobby-tables.com)
> image

**Comment by zspitz** [comment]
@petdance There are still some rendering bugs. What do you intend to do with them?

**Comment by petdance** [comment]
I'll fix 'em.  I just can't do it right now.

**Comment by petdance** [comment]
I've fixed all the rendering problems.  If there are any more problems you find, please make new issues for them, rather than add to this one.

**Comment by zspitz** [comment]
@petdance Should I also create new issues for content problems? Or should I create a new PR with corrections?

**Comment by petdance** [comment]
Yes, new issues for everything.  Let's count this PR as done.

**Comment by petdance** [comment]
Tweeted about this.  Thanks again @zspitz.  https://twitter.com/petdance/status/1032721124583514112

**Pull request closed without merging**